auth/azure/token_provider.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2024 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+this file except in compliance with the License.
+
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed
+under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+*/
+
+package azure
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+)
+
+const (
+	AzureDevOpsRestApiScope = "499b84ac-1321-427f-aa17-267ca6975798/.default"
+)
+
+// Provider is an authentication provider for Azure.
+type Provider struct {
+	credential azcore.TokenCredential
+	scopes     []string
+}
+
+// ProviderOptFunc enables specifying options for the provider.
+type ProviderOptFunc func(*Provider)
+
+// NewProvider returns a new authentication provider for Azure.
+func NewProvider(opts ...ProviderOptFunc) *Provider {
+	p := &Provider{}
+	for _, opt := range opts {
+		opt(p)
+	}
+	return p
+}
+
+// WithCredential configures the credential to use to fetch the resource manager
+// token.
+func WithCredential(cred azcore.TokenCredential) ProviderOptFunc {
+	return func(p *Provider) {
+		p.credential = cred
+	}
+}
+
+func WithAzureDevOpsScope() ProviderOptFunc {
+	return func(p *Provider) {
+		p.scopes = []string{AzureDevOpsRestApiScope}
+	}
+}
+
+func (p *Provider) GetToken(ctx context.Context) (*azcore.AccessToken, error) {
+	if len(p.scopes) == 0 {
+		return nil, fmt.Errorf("error scopes must be specified")
+	}
```

Comment:
I had thought about this while implementing the code. Using the azure public cloud scope does not work for Azure DevOps, but if we intend to share this package with ACR login for oci, it may be ok to default to azure public cloud and make sure that all git code with provider Azure overrides this to use the AzureDevOpsScope. I thought we can make the change for the default scope if/when we start using the package with oci. 

Please let me know your take on this and if you agree it is ok to default to azure public cloud scope in a subsequent PR when oci uses this package.